sklearn/metrics/cluster/_dbcv_internals.py

```diff
@@ -0,0 +1,316 @@
+        metrics, such as minkowski, Mahanalobis etc.
+
+    Returns
+    -------
+
+    distances : array (n_samples, n_samples)
+        The distances between all points in `X` with `label` equal to `cluster_id`.
+
+    core_distances : array (n_samples,)
+        The all-points-core_distance of all points in `X` with `label` equal
+        to `cluster_id`.
+
+    References
+    ----------
+    .. [1] `Moulavi, D., Jaskowiak, P.A., Campello, R.J., Zimek, A. and Sander, J.,
+    2014. Density-Based Clustering Validation. In SDM (pp. 839-847).
+    <https://www.dbs.ifi.lmu.de/~zimek/publications/SDM2014/DBCV.pdf>`_
+    """
+    if metric == "precomputed":
+        if d is None:
+            raise ValueError("If metric is precomputed a d value must be provided!")
+        distance_matrix = X[labels == cluster_id, :][:, labels == cluster_id]
+    else:
+        subset_X = X[labels == cluster_id, :]
+        distance_matrix = pairwise_distances(subset_X, metric=metric, **kwd_args)
+        if d is not None:
+            msg = (
+                'The "d" value you provided is being ignored. '
+                "It's only required for precomputed distances"
+            )
+            warnings.warn(msg, UserWarning)
+        d = X.shape[1]
+
+    if no_coredist:
+        return distance_matrix, None
+
+    else:
+        core_distances = _all_points_core_distance(distance_matrix.copy(), d=d)
+        core_dist_matrix = np.tile(core_distances, (core_distances.shape[0], 1))
+        stacked_distances = np.dstack(
+            [distance_matrix, core_dist_matrix, core_dist_matrix.T]
+        )
+
+        if print_max_raw_to_coredist_ratio:
+            print(
+                "Max raw distance to coredistance ratio: "
+                + str(_max_ratio(stacked_distances))
+            )
+
+        return stacked_distances.max(axis=-1), core_distances
+
+
+def _all_points_core_distance(distance_matrix, d=2):
+    """
+    Compute the all-points-core-distance for all the points of a cluster.
+
+    Parameters
+    ----------
+    distance_matrix : array (cluster_size, cluster_size)
+        The pairwise distance matrix between points in the cluster.
+
+    d : int
+        The dimension of the data set, which is used in the computation
+        of the all-point-core-distance as per the paper.
+
+    Returns
+    -------
+    core_distances : array (cluster_size,)
+        The all-points-core-distance of each point in the cluster
+
+    References
+    ----------
+    .. [1] `Moulavi, D., Jaskowiak, P.A., Campello, R.J., Zimek, A. and Sander, J.,
+    2014. Density-Based Clustering Validation. In SDM (pp. 839-847).
+    <https://www.dbs.ifi.lmu.de/~zimek/publications/SDM2014/DBCV.pdf>`_
+    """
+    non_diagonal_indices = ~np.eye(*distance_matrix.shape, dtype=bool)
+    if distance_matrix[non_diagonal_indices].min() == 0.0:
+        raise ValueError(
+            "Identified identical points in the input data. Such points cause their "
```

Comment:
Actually, I think it _can_ be undefined in the raw distance mode as well, since a raw dist of zero allows for the density separation between two clusters to be zero, if:
1. Two identical points are labeled with different labels.
2. Both points are considered internal nodes of their respective clusters.

The latter condition could feasibly happen at the fuzzy boundary of two clusters. My personal preference is to go even further and disallow identical points all-together (which would mean moving this check out of this function I suppose).

What do you think?

Reply:
Here are the options I see:

1. remove the new check, **disallow identical points all-together (single check)**
  - this would involve removing the current check, but it couldn't simply be moved a few abstractions "upwards" and has **heavy-ish implementation requirements**/implications (keep reading for details)
  - disregarding the details in terms of implementation, I'm not in support of this idea because it **severely limits practical applicability**, possibly to the point of irrelevance **for our users**, if we don't allow noise points whatsoever
  - **[Implementation considerations below, skip if you agree with my reasoning from user perspective anyway]**
  - for a thorough upfront duplication check we **would need to have access to a complete pairwise distance matrix for all the points in the given dataset**
    - at no point do we currently produce such a matrix, which is a behavior I'd like to preserve (even aside the impact on current abstraction boundaries) for the sake of memory needs (_the user can make use of the precomputed option though to pass such a matrix they produced themselves upfront_)
      - we only calculate pairwise intra-cluster distances for the density sparseness  [here](https://github.com/luis261/scikit-learn/blob/add-dbcv-metric/sklearn/metrics/cluster/_unsupervised.py#L600)
      - and then pairwise distances between points of 2 clusters at a time [here](https://github.com/luis261/scikit-learn/blob/add-dbcv-metric/sklearn/metrics/cluster/_unsupervised.py#L620)
      - this has big impact on space demand - which is obviously? the limiting factor when dealing with huge datasets, moreso than actual computing power in my experience
      - such a move from requiring no duplicate points per cluster to no duplicate points at all in the dataset would make such savings impossible if we go with/need the full matrix
        - we currently avoid the memory cost associated with storing the whole matrix at once (N^2: `(n_samples, n_samples)`), instead just taking on one cluster (or the points of two clusters paired with each other) at a time, which makes a large difference, even more so for higher cluster counts (and noise counts)
          - I'm simplifying here quite a bit, e.g. we do elect store the whole list of core distances accross iterations (but this is only `N`, not `N^2` in terms of space complexity) etc
  - on second thought, just **running a simple duplicate check based on raw coordinates instead of the calculated distances** would serve us just as well while circumventing the need for the whole matrix. However, this **would impact runtime while also being specific to the non-precomputed option.**
  -  **[Implementation considerations end]**
2. avoid undefined coredists as established, add a **dedicated check for failing on duplicates shared between clusters**
  - by just **leaving the established check as-is and opting for an additional duplicate check (including the diagonal this time) inside `_density_separation` instead, we can just make use of the matrix that gets subselected (for precomputed) or calculated anyway**
  - this does lead to a slight difference in terms of what we actually check for (although it shouldn't make a difference from a practical POV): we would now disallow duplicates inside a cluster for the MRD option only, just like before, and also now fail on duplicates present across at least two clusters, like you described
    - the nuance comes in because we would still allow duplicated points if they are either exclusively present in the noise or have a maximum occurence count of 1 accross all clusters
    - since we don't care about duplicate noise, this shouldn't be a problem, arguably an advantage instead
3. avoid undefined coredists as established, add a **dedicated check which raises only if the density separation between two clusters is zero**
- again, we would keep the new check specific for the MRD version, but supplement it with a separate check that specifically comes in just before the density separation is used in the denominator
- this option is the **least invasive, with neglibigle runtime cost overall, while still concisely avoiding the unintended problem we want to take care off**
- other than conciseness and ease of implementation, this also has the advantage of the check clause naturally directly preceding the actual line in terms of the location where the error could occur in the codepath
- as you can tell, I really like this option ..
4. avoid undefined coredists as established, add a dedicated check which warns only if the density separation between two clusters is zero
- **same as above, but we warn instead of raising**
- this would leave us more closely aligned with the scikit-learn contrib reference implementation, but allow diligent, educated users to opt-in to converting these warnings into errors on their own if they choose to do so
- **worth thinking about but probably not what we want**
  - disadvantage: in the case in question, a `ZeroDivisionError`/numpy runtime warning should show anyway, plus you can argue that something has definitely gone wrong if the density separation between two clusters is literally zero, to a point where the user should be informed about it with more than a warning?

**I propose we go with option 3.**

So I do think we can and should avoid the situation you describe, but without generally failing on duplicate input by adding an additional check, closer to where this other scenario could occur. So this would happen "downstream" instead of completely prohibting duplicates throughout the whole dataset upfront, which could be damaging for usability.